reference/solver/batch_gmres_kernels.cpp

```diff
@@ -0,0 +1,853 @@
+                ApplyPlaneRotation(
+                    &s_entry.values[inner_iter * s_entry.stride],
+                    &s_entry.values[(inner_iter + 1) * s_entry.stride],
+                    &cs_entry.values[inner_iter * cs_entry.stride],
+                    &sn_entry.values[inner_iter * sn_entry.stride], nrhs,
+                    converged);
+
+                ApplyPlaneRotation(
+                    &H_entry.values[inner_iter * H_entry.stride +
+                                    inner_iter * nrhs],
+                    &H_entry.values[(inner_iter + 1) * H_entry.stride +
+                                    inner_iter * nrhs],
+                    &cs_entry.values[inner_iter * cs_entry.stride],
+                    &sn_entry.values[inner_iter * sn_entry.stride], nrhs,
+                    converged);
+
+                for (int c = 0; c < nrhs; c++) {
+                    const uint32 conv = converged & (1 << c);
+
+                    if (conv) {
+                        continue;
+                    }
+                    H_entry.values[(inner_iter + 1) * H_entry.stride +
+                                   inner_iter * nrhs + c] = zero<ValueType>();
+                }
+
+
+                // estimate of residual norms
+                // residual = abs(s(inner_iter + 1))
+                for (int c = 0; c < res_norms_temp_entry.num_rhs; c++) {
+                    res_norms_temp_entry.values[c] = abs(
+                        s_entry.values[(inner_iter + 1) * s_entry.stride + c]);
+                }
+
+                copy(gko::batch::to_const(res_norms_temp_entry),
+                     res_norms_entry, converged);
+
+                const uint32 converged_prev = converged;
+
+                all_converged = stop.check_converged(
+                    outer_iter * (restart + 1) + inner_iter + 1,
+                    res_norms_entry.values, {NULL, 0, 0, 0}, converged);
+
+                const uint32 converged_recent = converged_prev ^ converged;
+
+                // y = Hessenburg(0 : inner_iter,0 : inner_iter) \ s(0 :
+                // inner_iter) x = x + KrylovBasis(:, 0 : inner_iter ) * y
+                update_x(x_entry, inner_iter, gko::batch::to_const(H_entry),
+                         gko::batch::to_const(s_entry),
+                         gko::batch::to_const(V_entry), ~converged_recent);
+
+                logger.log_iteration(
+                    ibatch, outer_iter * (restart + 1) + inner_iter + 1,
+                    res_norms_entry.values, converged);
+
+                if (all_converged) {
+                    inner_loop_break_flag = true;
+                    break;
+                }
+            }
+
+            if (inner_loop_break_flag == true) {
+                break;
+            }
+
+            // y = Hessenburg(0:restart - 1,0:restart - 1) \ s(0:restart-1)
+            // x = x + KrylovBasis(:,0 : restart - 1) * y
+            update_x(x_entry, restart - 1, gko::batch::to_const(H_entry),
```

Comment:
Here, it might be better to separate the triangular solve from the update of `x`. Right now, at the end of every restart, it looks like you're solving the triangular system again for no reason.

Reply:
This is not very clear to me, why don't we want the triangular solve at the end of every restart? Isn't that necessary for the update of x?